File: src/main/java/com/samla/sdk/Samla.java
package com.samla.sdk;

import android.app.Activity;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentActivity;
import androidx.fragment.app.FragmentManager;
import android.util.Log;
import android.view.View;

import androidx.lifecycle.Lifecycle;
import androidx.lifecycle.LifecycleObserver;
import androidx.lifecycle.LifecycleOwner;
import androidx.lifecycle.OnLifecycleEvent;

import com.google.firebase.FirebaseApp;
import com.google.firebase.FirebaseAppLifecycleListener;
import com.google.firebase.FirebaseOptions;
import com.google.firebase.analytics.FirebaseAnalytics;
import com.samla.sdk.analytics.ActivityManager;
import com.samla.sdk.analytics.ViewManager;
import com.samla.sdk.analytics.funnel.FunnelManager;
import com.samla.sdk.userinterface.UserInterfaceHierarchy;

import static androidx.lifecycle.Lifecycle.Event.*;

public class Samla implements LifecycleObserver, SamlaBuilder, FragmentManager.OnBackStackChangedListener {
    private final static String TAG = Samla.class.getSimpleName();

    private Activity applicationActivity;
    private Lifecycle lifecycle;
    private FunnelManager funnelManager;
    private ActivityManager activityManager;
    private static boolean userFlowCreationEnabled = true;

    FirebaseAnalytics firebaseAnalytics;
    FirebaseAppLifecycleListener firebaseAppLifecycleListener;
    FirebaseApp firebaseApp;
    FirebaseOptions firebaseOptions;

    public Samla(Activity activity) {
        this.applicationActivity = activity;
        funnelManager = new FunnelManager(applicationActivity);
        activityManager = new ActivityManager(applicationActivity);
        //applicationActivity.addOnBackStackChangedListener(this);
    }

    public static Samla withActivity(Activity activity) {
        return new Samla(activity);
    }

    /**
     * A userflow is a visual overview of an application's screens/navigation and structure
     * @param enable: Enables/disables the creation of a userflow when building the application
     */
    public static void createUserFlow (Boolean enable) {
           userFlowCreationEnabled = enable;
    }

    @Override
    public SamlaBuilder withLifeCycle(Lifecycle lifeCycle) {
        this.lifecycle = lifeCycle;
        this.lifecycle.addObserver(this);
        return this;
    }

    public Activity getApplicationActivity() {
        return applicationActivity;
    }

    public FunnelManager getFunnelManager() {
        return funnelManager;
    }

    @OnLifecycleEvent(ON_CREATE)
    void onCreated(LifecycleOwner lifecycleOwner) {
        Log.i(TAG, "onCreated");
        UserInterfaceHierarchy.setMenuListener(applicationActivity.findViewById(android.R.id.content));
    }

    @OnLifecycleEvent(ON_RESUME)
    void onResume() {
        Log.i(TAG, "onResume");
    }

    @OnLifecycleEvent(ON_PAUSE)
    void onPause () {
        Log.i(TAG, "onPause");
    }

    @OnLifecycleEvent(ON_STOP)
    void onStop () {
        Log.i(TAG, "onStop");

        String hierarchy = UserInterfaceHierarchy.logViewHierarchy(applicationActivity);
        Log.i(TAG, "UI Hierarchy: " + hierarchy);

        UserInterfaceHierarchy.takeScreenShot(applicationActivity);

        if (userFlowCreationEnabled) {
            // Todo: Testing to send
        }

    }

    @OnLifecycleEvent(ON_DESTROY)
    void onDestroy () {
        Log.i(TAG, "onDestroy");
    }

    @OnLifecycleEvent(ON_ANY)
    void onAny(LifecycleOwner source, Lifecycle.Event event) {

    }

    void test () {
        FragmentManager manager = new FragmentManager() {
            @Nullable
            @Override
            public Fragment findFragmentByTag(@Nullable String tag) {
                return super.findFragmentByTag(tag);
            }

            @Override
            public void addOnBackStackChangedListener(@NonNull OnBackStackChangedListener listener) {
                super.addOnBackStackChangedListener(listener);
            }

            @Override
            public void removeOnBackStackChangedListener(@NonNull OnBackStackChangedListener listener) {
                super.removeOnBackStackChangedListener(listener);
            }
        };
        manager.findFragmentByTag("SimpleClassName");


    }

    @Override
    public void onBackStackChanged() {
        Log.i(TAG, "onBackStackChanged");
    }
}

File: src/main/java/com/samla/sdk/Samla.kt
package com.samla.sdk

import android.app.Activity
import android.content.Context
import android.util.Log
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleObserver
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.OnLifecycleEvent
import com.google.firebase.FirebaseApp
import com.google.firebase.FirebaseAppLifecycleListener
import com.google.firebase.FirebaseOptions
import com.google.firebase.analytics.FirebaseAnalytics
import com.samla.sdk.storage.DataStorage
import com.samla.sdk.userflow.Analytics
import com.samla.sdk.userflow.funnel.FunnelManager
import com.samla.sdk.userinterface.ActivityManager
import com.samla.sdk.userinterface.UIHierarchy


class Samla constructor(context : Context) : LifecycleObserver, SamlaBuilder, FragmentManager.OnBackStackChangedListener, ISamla {
    private val TAG = Samla::class.java.simpleName

    val mContext : Context
    val mActivity: Activity
    lateinit var mLifeCycle: Lifecycle;

    lateinit var firebaseAnalytics: FirebaseAnalytics;
    lateinit var firebaseAppLifecycleListener: FirebaseAppLifecycleListener;
    lateinit var firebaseApp: FirebaseApp;
    lateinit var firebaseOptions: FirebaseOptions;


    init {
        mContext  = context
        mActivity = context as Activity

        FunnelManager.setClientActivity(this);
        ActivityManager.setClientActivity(this);
        DataStorage.setClientActivity(this);
        Analytics.setClientActivity(this);
    }

    override fun getActivity(): Activity {
        return mActivity
    }

    override fun withLifeCycle(lifeCycle: Lifecycle): SamlaBuilder {
        mLifeCycle = lifeCycle
        mLifeCycle.addObserver(this)
        return this
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_CREATE)
    fun onCreate(lifecycleOwner: LifecycleOwner?) {
        Log.i(TAG, "onCreated")

    }

    @OnLifecycleEvent(Lifecycle.Event.ON_RESUME)
    fun onResume() {
        Log.i(TAG, "onResume")
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_PAUSE)
    fun onPause() {
        Log.i(TAG, "onPause")
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_STOP)
    fun onStop() {
        Log.i(TAG, "onStop")
        val hierarchy = UIHierarchy.logViewHierarchy(mActivity)
        Log.i(TAG, "UI Hierarchy: $hierarchy")
        UIHierarchy.getScreenshot(mActivity)
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_DESTROY)
    fun onDestroy() {
        Log.i(TAG, "onDestroy")
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_ANY)
    fun onAny(source: LifecycleOwner?, event: Lifecycle.Event?) {
    }

    fun menuListener () {
        UIHierarchy.setMenuListener(
            mActivity.findViewById(
                R.id.content
            )
        )
    }

    fun test() {
        val manager: FragmentManager =
            object : FragmentManager() {
                override fun findFragmentByTag(tag: String?): Fragment? {
                    return super.findFragmentByTag(tag)
                }

                override fun addOnBackStackChangedListener(listener: OnBackStackChangedListener) {
                    super.addOnBackStackChangedListener(listener)
                }

                override fun removeOnBackStackChangedListener(listener: OnBackStackChangedListener) {
                    super.removeOnBackStackChangedListener(listener)
                }
            }
        manager.findFragmentByTag("SimpleClassName")
    }

    override fun onBackStackChanged() {
        Log.i(TAG, "onBackStackChanged")
    }

    /**
     * A userflow is a visual overview of an application's screens/navigation and structure
     * @param enable: Enables/disables the creation of a userflow when building the application
     */
    fun createUserFlow(enable: Boolean) {

    }

}